T196 fix additional frame

Funth0mas opened the pull request:
Extend `--additionalFrameAttributes` to support any attribute name. Add this feature to json export. Replace `eval()` by `getattr()`.

altendky commented:
Yeah, `eval()` is...  ugly.  But, I'm not a huge fan of making dict keys available as attributes either.  The cli effect could be achieved without that as well.  I can't say I have a really strong argument, it mostly just seems less 'magical'.  (and yes, I have worked with `__getattr__` etc, so I don't mean magical like I am not familiar with it)

```python
getattr(signal, item, signal.attributes.get(item, ""))
```

Or just add a getter.

```python
NOTHING = object()
def get(self, item, default=NOTHING):
    item = getattr(self, item, self.attributes.get(item, default))

    if item is NOTHING:
        raise SomeException()

    return item
```
```python
signal.get(item, "")
```

Funth0mas commented:
Totally agree, your solution is better. I'll apply it.

There is still question if it make sense to have some "additional attributes" as instance variables and some in the attributes dictionary. It would make sense to have them all on a single place. But I didn't want to touch the existing code, because...  I'm just a passerby.

altendky commented:
If I remember a conversation with @ebroecker correctly, it would be nice to revisit these 'attributes' and try to figure out which are common and how to nestle the rest together happily.  Basically, there is a dbc bias and with the extra knowledge of other formats we could probably do a better job picking regular attributes to have.

Thanks for seeing the mess and doing something about at least part of it.  :]  We appreciate the help.

ebroecker commented:
@Funth0mas 
having all attributes in one place is not that easy...
My intention was to have the "common" attributes direct in the frame (yes some are missing). 
The attributes[]-Dictionary is more or less a "dbc-only" place to save all dbc-attributes. The problem of dbc is, that you can define any attribute in dbc. 
Some attributes are not yet in the common list (#146)

altendky commented on src/canmatrix/canmatrix.py:
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
+        :param default: Default value if attribute doesn't exist.
+        :return any: Return the attribute value if found, else `default` or None
+        """
+        if hasattr(self, name):
```
Sorry to evolve my request here...  I forgot this was an `attrs` class and I think this will be a touch better.  It will avoid considering methods and such as attributes.

https://www.attrs.org/en/stable/api.html#attr.fields_dict

```python
if name in attr.fields_dict(type(self)):
```

I'll note that this is technically a breaking change given the it will block access via this method to `.attribute` items that shadow the names of regular attributes.  But, I don't think that bothers me.

altendky commented on src/canmatrix/canmatrix.py:
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
+    #       self._sendType = value
+    #       self.attributes["GenMsgCycleTime"] = value
+
+    def attribute(self, name, db=None, default=None):
```
This is definitely a breaking change though it does seem better.  I think it's better for `canmatrix` to be tending towards improving so I'm comfortable with this but figured I'd highlight it.

Funth0mas replied:
If this change is accepted, I'd adopt the similar functions of another classes accordingly.

ebroecker replied:
canmatrix is at 0. - version. I agree that improving is always better than staying compatible.  
I'd accept it!

Funth0mas replied:
I'll refactor the `attribute()` of another classes to have the same signature in another PR (when I don't fall asleep in the train next time)

altendky commented on src/canmatrix/canmatrix.py:
```diff
@@ -425,62 +436,75 @@ def pgn(self, value):
                 return define.defaultValue
         else:
-            return None
+            return default
```
+1

altendky commented on src/canmatrix/cmjson.py:
```diff
@@ -101,21 +101,21 @@ def dump(db, f, **options):
             for frame_info in additionalFrameColums:  # Look for additional Frame Attributes
-                if hasattr(frame, frame_info):
-                    frame_attributes[frame_info] = getattr(frame, frame_info)
+                if frame.attribute(frame_info):  # does the attribute exist?
```
This doesn't check it's existence.  If it exists but is `None` or anything else falsey it will be treated as not present.  Here's the first thing that came to my mind, though it is a touch smelly to me.  Maybe you can think of something nicer.

```python
MISSING = object()
value = frame.attribute(name=frame_info, default=MISSING)
if value is not MISSING:
    frame_attributes[frame_info] = value
```

Funth0mas replied:
good point! As there is imho no reason to export None value, change to `is not None`

altendky commented on src/canmatrix/cmjson.py:
```diff
@@ -101,21 +101,21 @@ def dump(db, f, **options):
             frameattribs = {}
             for attribute in db.frameDefines:
-                frameattribs[attribute] = frame.attribute(db,attribute)
+                frameattribs[attribute] = frame.attribute(attribute, db)
```
Let's specify by name to avoid this hazard in the future.  Looks like this applies to most of the rest of the diff below.

Funth0mas replied:
Uff, I realized the PyCharm has refactored also calls of another classes. Python has certain weaknesses :-(
Checked, reverted unwanted changes. For Frame use named attribute.

Funth0mas commented:
I ran the autoformating on the whole file by accident. If it's problem, I'll revert it line per line. Sorry for the inconvenience. 
Mainly I removed the cryptic getattr and I propose to use the existing attribute() getter, but make the db parameter really optional. I added some docstrings (only in Frame) when trying to understand the code.

altendky commented:
No big deal, I just tend to mention more rather than less.  Thanks for the help improving the code.

Funth0mas commented on src/canmatrix/canmatrix.py:
```diff
@@ -370,45 +374,52 @@ def __str__(self):
-    comment
+    Contains one CAN Frame.
+
+    The Frame has  following mandatory attributes\n
```
There must be either empty line or \n befor the list (ctrl+Q in PyCharm). Changed to empty line.

Funth0mas commented on src/canmatrix/canmatrix.py:
```diff
@@ -105,49 +106,50 @@ class Signal(object):
-    name = attr.ib(default = "")
-#    float_factory = attr.ib(default=defaultFloatFactory)
+    name = attr.ib(default="")
+    #    float_factory = attr.ib(default=defaultFloatFactory)
```
(comic is that it was done automatically by PyCharm) ... clear, reverted, the same for all another unwanted changes. Comments, indentation and another codestyle issues can be fixed later in own PR.

Funth0mas commented on src/canmatrix/canmatrix.py:
```diff
@@ -780,7 +816,8 @@ def __init__(self, definition):
             if out != defaultFloatFactory(inStr):
-                logger.warning("Warning, integer was expected but got float: got: {0} using {1}\n".format(inStr, str(out)))
+                logger.warning(
+                    "Warning, integer was expected but got float: got: {0} using {1}\n".format(inStr, str(out)))
```
unwanted change, revert.

Funth0mas commented:
Ready for review again. Thanks!

ebroecker commented on src/canmatrix/canmatrix.py:
```diff
@@ -409,6 +416,7 @@ class Frame(object):
 
     @property
     def is_multiplexed(self):
+        """Frame is multiplexed if at least one of its signals is multiplexed."""
```
Frame is multiplexed it at least one of its signals is a multiplexer

ebroecker commented on src/canmatrix/canmatrix.py:
```diff
@@ -552,45 +584,54 @@ def globSignals(self, globStr):
         try:
             self.attributes[attribute] = str(value)
-        except:
+        except:  # could it really fail?
```
I'm not sure why I added this try-catch. But I'm quite sure that there was a problem without it.
Maybe there is a problem with "str" if your are in python2 and have some issues with charsets...

altendky replied:
None of this should be dealt with in this PR but:

1) it should catch the particular exception not just anything
2) i think users should be expected to pass the proper type
3) i think there are still various str-or-int-or-... around so maybe it would get caught up in cleaning that up.

ebroecker commented on src/canmatrix/cmcsv.py:
```diff
@@ -183,10 +183,7 @@ def dump(db, thefile, delimiter=',', **options):
-                temp = eval("frame." + frameInfo)
-            except:
-                temp = ""
+            temp = frame.attribute(frameInfo, default="")
```
+++ this eliminates my dirty "eval"-hack.
Thanks

altendky commented on src/canmatrix/cmcsv.py:
```diff
@@ -270,3 +261,7 @@ def dump(db, thefile, delimiter=',', **options):
+    if sys.version_info > (3, 0):
+        # When TextIOWrapper is garbage collected, it closes the raw stream
+        # unless the raw stream is detached first
+        temp.detach()
```
Funny thing, I just ran into this over in a tqdm PR.  Sensible addition given the `if` above but the thing overall sort of smells to me.  `*hopes he wasn't the one that added it*`

Should probably really be in a finally clause though just in case there are exceptions.  Or a context manager, whatever you feel like.

Funth0mas replied:
I simply copied it from another module because there were an error/warning after running the Unit Test informing about unreleased resource. To be honest I don't know why the `TextIOWrapper` must be present. Either let it be for further cleanup (I can create a ticket) or I need more time to investigate. The first solution is currently more convenient for me :-)

altendky replied:
Yeah, all I mean for this is to make a try/finally to be sure the detach actually gets run even in case of exceptions.

Funth0mas commented:
The failed appveyor isn't my fault  :-(

Funth0mas commented:
Hi, I merged the current origin/development and now it seems I did more changes than I really did, when looking to "changed files". I'm used to do this with bitbucket but it has different result. Did I something wrong?

altendky commented:
@Funth0mas are you referring to 3bdafa9?  I generally do a `git merge --no-ff development` and get a merge commit.  I'm not really sure quite why it was diffing the way it was...  but I think I just fixed it.  Fwiw, I did a `git rebase -i HEAD~3` and deleted the squash commit then did a regular `git merge development --no-ff` and a force push to your branch (you let us have write access to just that branch when you submitted the PR).  Yes, I did also push a backup branch to my repo first.  :]

Anyways, if this looks right to you I'm ready to merge.

Funth0mas commented:
@altendky funny thing is that I did almost the same except of using more complicated `git reset` and force pushed two minutes before you did :-) At least I understand the git a little bit more now. 
I 'inspected' the diff once again and yes, it's the state I wanted to have. Thank you for your attention.

altendky commented:
Thanks for all the code and your responsiveness.

Cheers,
-kyle

Funth0mas commented:
My pleasure. BTW you're from Germany?
Regards

ebroecker commented:
No, only I am from Germany, altendkey Not...
